spinalcordtoolbox/registration/algorithms.py

```diff
@@ -329,6 +337,220 @@ def register_step_label(src, dest, step, verbose=1):
+        if i < i_max:
+            torchparam[k] = weights_second_model[i]
+            i += 1
+    pt_second_model.load_state_dict(torchparam)
+    pt_second_model.to(device)
+    pt_second_model.eval()
+
+    # ---- Registration ---- #
+    logger.info("\n Predicting using first VxmDense model...")
+    moved, warp_tensor_first = pt_first_model(input_moving, input_fixed, registration=True)
+    warp_data_first = warp_tensor_first[0].permute(1, 2, 3, 0).detach().numpy()
+
+    logger.info("\n Predicting using second VxmDense model...")
+    moved_final, warp_tensor_second = pt_second_model(moved, input_fixed, registration=True)
+    warp_data_second = warp_tensor_second[0].permute(1, 2, 3, 0).detach().numpy()
+
+    # ---- Warping fields ---- #
+    # Modify the warp data so it can be used with sct_apply_transfo()
+    # (add a time dimension, change the sign of some axes and set the intent code to vector)
+    orientation_conv = "LPS"
+    fx_im_orientation = list(aff2axcodes(fx_preproc_nii.affine))
+    opposite_character = {'L': 'R', 'R': 'L',
+                          'A': 'P', 'P': 'A',
+                          'I': 'S', 'S': 'I'}
+    perm = [0, 1, 2]
+    inversion = [1, 1, 1]
+    for i, character in enumerate(orientation_conv):
+        try:
+            perm[i] = fx_im_orientation.index(character)
+        except ValueError:
+            perm[i] = fx_im_orientation.index(opposite_character[character])
+            inversion[i] = -1
+
+    # First Warping Field
+    # Add the time dimension
+    warp_data_exp_first = np.expand_dims(warp_data_first, axis=3)
+    warp_data_exp_first_copy = np.copy(warp_data_exp_first)
+    # Permute the axes if needed
+    warp_data_exp_first[..., 0] = inversion[0] * warp_data_exp_first_copy[..., perm[0]]
+    warp_data_exp_first[..., 1] = inversion[1] * warp_data_exp_first_copy[..., perm[1]]
+    warp_data_exp_first[..., 2] = inversion[2] * warp_data_exp_first_copy[..., perm[2]]
+
+    # Second Warping Field
+    # Add the time dimension
+    warp_data_exp_second = np.expand_dims(warp_data_second, axis=3)
+    warp_data_exp_second_copy = np.copy(warp_data_exp_second)
+    # Permute the axes if needed
+    warp_data_exp_second[..., 0] = inversion[0] * warp_data_exp_second_copy[..., perm[0]]
+    warp_data_exp_second[..., 1] = inversion[1] * warp_data_exp_second_copy[..., perm[1]]
+    warp_data_exp_second[..., 2] = inversion[2] * warp_data_exp_second_copy[..., perm[2]]
+
+    # Compose the two warping fields together (from the first and second registration)
+
+    # Forward registration
+    warp_data = warp_data_exp_first + warp_data_exp_second
+    warp = Nifti1Image(warp_data, fx_preproc_nii.affine)
+    # Set the intent code to vector
+    warp.header['intent_code'] = 1007
+    # Save the composed warping field [forward]
+    save(warp, fname_warp_forward)
+
+    # Reverse registration
+    warp_data_rev = -warp_data  # TODO - observe if it provides good results
```

Comment:
Adding a note for this TODO here.

Reply:
@EvanBeal Do you think this TODO is still needed? Has this worked the way you expected it to?

Reply:
Yes it seems to work as expected so I have removed it now.